compiler-rt/lib/builtins/CMakeLists.txt

```
@@ -756,8 +756,15 @@ set(riscv64_SOURCES
   ${riscv_SOURCES}
 )
 
-set(sparc_SOURCES ${GENERIC_SOURCES} ${GENERIC_TF_SOURCES})
-set(sparcv9_SOURCES ${GENERIC_SOURCES} ${GENERIC_TF_SOURCES})
+# Exclude the FT sources for 32-bit SPARC.  Clang doesn't
+# support 128-bit long double on 32-bit SPARC.
+if("${COMPILER_RT_DEFAULT_TARGET_ARCH}" MATCHES "sparc")
+  set(sparc_SOURCES ${GENERIC_SOURCES})
+  set(sparcv9_SOURCES ${GENERIC_SOURCES})
+else()
+  set(sparc_SOURCES ${GENERIC_SOURCES} ${GENERIC_TF_SOURCES})
+  set(sparcv9_SOURCES ${GENERIC_SOURCES} ${GENERIC_TF_SOURCES})
+endif()
```

I don't quite understand this conditional. Why do we need to check `COMPILER_RT_DEFAULT_TARGET_ARCH`?

Also is the sparcv9 build 32-bit? If not could this be simplified to:
```suggestion
set(sparc_SOURCES ${GENERIC_SOURCES})
set(sparcv9_SOURCES ${GENERIC_SOURCES} ${GENERIC_TF_SOURCES})
```

---

This is something I need @rorth feedback on.  I don't know enough about the versions of sparc and what's supported or not.  From the discussions, it sounded like "sparc" is 32-bit and the other `sparc*` are 64-bit.  

I think you are right.  Hopefully Rainer, provides feedback too.

---

ping @rorth . I think this is the only open issue blocking this from being merged.

---

I've meanwhile tried 2-stage builds on both `sparcv9-sun-solaris2.11` and `sparc64-unknown-linux-gnu`, each starting from `gcc-14` and `clang-19`.  There were no failures.

However, the above is as wrong as it can get: on both Solaris/sparcv9 and Linux/sparc64 there were no `TF` files
compiled any longer, neither `sparc` nor `sparcv9`.  For one, AFAICS `MATCHES` isn't anchored, so it matches all of `sparc`, `sparcv9`, and 
`sparc64`.  Besides, the concrete default  SPARC target arch is completely irrelevant here: both a 64-bit-default compiler and a
32-bit-default compiler can produce both 32 and 64-bit objects.

---

Not right now.  I seriously think we need to take two steps back first.  So far, you haven't provided some quite  crucial information:
- What is the goal of this patch?
- What exactly fails with current `main`?
- Then for every change we need a detailed justification why the current code is wrong, while your proposed change is right, not just for s390x, but in general.

Right now, I can't help but feel that we've poking in the dark for months without any real progress or even understanding what the problem is.  This is extremely tiring: I at least have long run out of energy, time, and patience about this patch.  The upcoming releases of binutils 2.44, GCC 15, and LLVM 20 don't help in the slightest for this.

An added problem is the fact that to the best of my knowledge there's no public s390x system where one could try things for your target.  OTOH, there's a Solaris/sparcv9 system in the cfarm that is perfectly capable of building and testing LLVM.  I don't even know if there's any s390x ABI document available, so knowing the system's properties is all but impossible.

---

>Then for every change we need a detailed justification why the current code is wrong, while your proposed change is right, not just for s390x, but in general.

@arichardson has reviewed this and approves the change.  In this case the changes recently made to divtc3.c and multc3.c had a bad assumption connecting native 128-bit long double support to 128-bit int support.  That is not generally correct.  I could simply reverse that but I know that will break sparc hence the change to this file.

The answers to the other questions are in previous PRs.  I'll add the answers in the description of this one to save searching.